zebrad/src/commands/start/sync.rs

```diff
@@ -284,15 +284,35 @@ where
                 match res.map_err::<Report, _>(|e| eyre!(e)) {
                     Ok(zn::Response::BlockHashes(hashes)) => {
                         tracing::debug!(first = ?hashes.first(), len = ?hashes.len());
+                        tracing::trace!(?hashes);
 
+                        // zcashd sometimes appends an unrelated hash at the
+                        // start or end of its response. Check the first hash
+                        // against the previous response, and discard mismatches
                         let unknown_hashes = match hashes.split_first() {
                             None => continue,
                             Some((expected_hash, rest)) if expected_hash == &tip.expected_next => {
                                 rest
                             }
-                            Some((other_hash, _rest)) => {
-                                tracing::debug!(?other_hash, ?tip.expected_next, "discarding response with unexpected next hash");
-                                continue;
+                            Some((other_hash, rest)) => {
+                                // See if it's just one extra hash
+                                // TODO: un-nest matches, probably by extracting this logic into a function
+                                match rest.split_first() {
```

Comment:
I tried to refactor the code to do a single match, but the `rest` slice is a different depending on whether the first or second hash matches. So I would have needed first, first_rest, second, and second_rest.

At that point, it was more complicated than the nesting.

Reply:
I think rather than using `split_first` you might be able to use slice patterns to handle this

Something like...:

```rust
                        let unknown_hashes = match hashes.deref() {
                            [] => continue,
                            [only_hash] if only_hash == &tip.expected_next => todo!(),
                            [first, second, rest @ ..] => todo!(),
```